[ZIPs 207, 214, 251] Consensus ZIPs for Zcash Development Fund

daira opened the pull request:
closes #328 and #330

daira committed:
ZIP 207: line wrapping (to reduce the subsequent diff).

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
Reintroduce ZIP 207.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
Add ZIPs 214 and 251.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

str4d commented on zip-0214.rst:
```diff
@@ -0,0 +1,327 @@
+==========
+
+.. [#RFC2119] `Key words for use in RFCs to Indicate Requirement Levels <https://www.rfc-editor.org/rfc/rfc2119.html>`_
+.. [#protocol] `Zcash Protocol Specification [Overwinter+Sapling+Blossom] <protocol/protocol.pdf>`_
```
Missing version.

str4d commented on zip-0251.rst:
```diff
@@ -0,0 +1,134 @@
+==========
+
+.. [#RFC2119] `Key words for use in RFCs to Indicate Requirement Levels <https://tools.ietf.org/html/rfc2119>`_
+.. [#protocol] `Zcash Protocol Specification, Version 2019.0.4 or later [Overwinter+Sapling+Blossom] <protocol/protocol.pdf>`_
```
Make sure this is the version that was referenced when writing the ZIP.

daira committed:
ZIP 251: Add wording about transaction digest algorithm and reference ZIP 243.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 214: Update version in Protocol Specification reference.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 251: remove trailing whitespace.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 207: s/founders reward/Founders' Reward/

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira commented on zip-0214.rst:
```diff
@@ -0,0 +1,327 @@
+
+The term "network upgrade" in this document is to be interpreted as
+described in ZIP 200 [#zip-0200]_ and the Zcash Trademark Donation and License
+Agreement ([#trademark]_ or successor agreement).
```
Here's the wording in the Zcash Trademark Donation and License Agreement:

> 1.9 “Zcash” means: (a) the digital currency whose block at height 0 has the hash
00040fe8ec8471911baa1db1266ea15dd06b4a8a5c453883c000b031973dce08 and whose block at height 419200 has the hash 00000000025a57200d898ac7f21e26bf29028bbe96ec46e05b2c17cc9db9e4f3 (the “Reference
Implementation”); or (b) any subsequent upgrade or other change to the Reference Implementation (“Network Upgrade”) agreed upon by the parties under Section 6.2(b)(i).

where Section 6.2(b)(i) says:

> If the parties each agree that the Network Upgrade should become the new
Reference Implementation of Zcash, the parties will take the actions specified in Exhibit D (Network Upgrades) to memorialize and implement their agreement.

If I understand correctly, this uses "Network Upgrade" in a way that is intended to include any deliberate consensus change on Mainnet. All other Zcash documentation uses "network upgrade" only for upgrades (on Mainnet or Testnet) using the mechanism defined in [ZIP 200](https://zips.z.cash/zip-0200). For instance, the changes implemented for zcashd in #3885 or #4331 were consensus changes, on both Mainnet and Testnet, that were not network upgrades.

This is somewhat annoying to fix without changing the Agreement or ZIP 1014. How should we proceed?

daira replied:
Oh, it actually doesn't matter because the only uses of "network upgrade" in ZIP 214 are unambiguous in context.

daira committed:
ZIP 214 is applicable only to Zcash.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 207: add definitions of "branch" and "network upgrade".

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 214: state explicitly that start height is inclusive and end height is exclusive.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 214: change wording to reflect the fact that ZEC != TAZ.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 214: add rationale section.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira commented:
@str4d all your comments should have been addressed, please re-review.

We also need a reviewer from ZF.

daira committed:
"branch" -> "consensus branch"

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

gtank reviewed (commented):
This is an artifact of ECC's key management policy, not consensus necessity. It's likely that ZF and notional direct grant recipients may have different address handling. We need to think about this.

daira committed:
Scrunch pre and math blocks a bit more.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 207: simplify wording about using a proportion of block subsidy.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 207: use math markup.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 207: in the example implementation, ensure that multiplication by the numerator won't overflow.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 207: be more explicit about which consensus rule is no longer active after NU4 activation.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 207: reference fix and HTML.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira commented:
Waiting on @acityinohio to decide whether ZF objects to the requirement for independent addresses.

daira commented on zip-0214.rst:
```diff
@@ -0,0 +1,365 @@
+For each of Testnet and Mainnet, before deploying this ZIP in a node implementation
+with the activation height set for that network, each of the parties (ECC and ZF)
+SHALL generate sequences of independent recipient addresses to be used for each
+stream in each funding period:
```
"Independent" was intended to mean that each address is generated independently at random, yes.

acityinohio commented:
> Waiting on @acityinohio to decide whether ZF objects to the SHALL requirement for independent addresses.

Thanks for the mention @daira; I agree with @gtank and based on the discussion here I'd definitely prefer SHOULD instead of SHALL; it would give us more leeway given the Foundation's current custody arrangements, which lend themselves to using the same address. I'm perhaps not the best person to talk about the trade-offs here but I can say that we feel comfortable using the same address for the time being, given the environment in which it was generated, the insurance against it, and how it's stored.

daira committed:
Protocol spec: clarify note about hashFinalSaplingRoot/hashLightClientRoot.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
Protocol spec: add references for the NU4 upgrade.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIP 214: blockchains -> block chains.

Co-Authored-By: str4d <thestr4d@gmail.com>

daira committed:
ZIP 214: allow addresses within a funding stream to be repeated.
Remove definitions of unused RFC 2119 keywords.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

defuse commented on zip-0207.rst:
```diff
@@ -434,7 +349,8 @@ Example implementation
-                    return state.DoS(100, error("%s: founders reward missing", __func__), REJECT_INVALID, "cb-no-founders-reward");
+                    return state.DoS(100, error("%s: founders reward missing", __func__),
+                                     REJECT_INVALID, "cb-no-founders-reward");
                 }
```
How much more effort would it take to make this code actually runnable with some tests?

daira replied:
When we write the implementation PR, we can update the ZIP with whatever code is actually used.

defuse commented on zip-0207.rst:
```diff
@@ -335,14 +254,9 @@ Example implementation
 
```
Above the fold the code sets the `startHeight` and `endHeight` for all funding streams to the wrong values, starting at blossom activation and ending at the first halving. It should be made consistent with ZIP 214.

daira commented on zip-0207.rst:
```diff
@@ -43,243 +80,129 @@ Specification
+        \mathsf{AddressChangeInterval} &=& \mathsf{PostBlossomHalvingInterval} / 48 \\
+        \mathsf{AddressPeriod}(\mathsf{height}) &=&
+            \mathsf{floor}\left(
+                {\small\frac{\mathsf{height} + \mathsf{PostBlossomHalvingInterval} - \mathsf{HeightForHalving}(1)}{\mathsf{AddressChangeInterval}}}
```
Yeah I don't see why that's necessary. It doesn't make any difference because only differences between two outputs of AddressPeriod are used in the computation of FundingStream[Fund].AddressIndex(height). I guess a C++ implementation would have to be more careful to implement `floor` correctly, since without the addition of PostBlossomHalvingInterval, the AddressPeriod could be negative before the first halving. (This cannot happen on mainnet with the planned activation height, but can happen on testnet.)

daira replied:
@str4d, do you think we should change this, or not?

daira commented on zip-0207.rst:
```diff
@@ -295,14 +218,9 @@ Example implementation
     };
 
```
Block heights are represented as `int` everywhere in the `zcashd` C++ source.

daira committed:
ZIP 207: improvements to the pseudocode.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira committed:
ZIPs 208, 250 and 251: keep `NETWORK_UPGRADE_PEER_PREFERENCE_BLOCK_PERIOD` the same
in terms of blocks, but it represents 1.5 days for the Heartwood upgrade onward.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>

daira commented:
This still needs an ACK from a ZF reviewer.

daira committed:
ZIP 251: update versions of zcashd that are planned to support NU4.

Signed-off-by: Daira Hopwood <daira@jacaranda.org>
